File: lib/models/product.dart
import 'category.dart';

class Product {
  final int id;
  final int categoryId;
  final String name;
  final String description;
  final double price;
  final int stock;
  final String? image;
  final bool isActive;
  final Category? category;

  Product({
    required this.id,
    required this.categoryId,
    required this.name,
    required this.description,
    required this.price,
    required this.stock,
    this.isActive = true,
    this.image,
    this.category,
  });

  factory Product.fromJson(Map<String, dynamic> json) {
    return Product(
      id: json['id'],
      categoryId: json['category_id'] ?? 0,
      name: json['name'] ?? 'Unknown',
      description: json['description'] ?? '',
      price: double.tryParse(json['price'].toString()) ?? 0.0,
      stock: int.tryParse(json['stock'].toString()) ?? 0,
      isActive: json['is_active'] == 1 || json['is_active'] == true || json['is_active'] == "1",
      image: json['image'],
      category: json['category'] != null ? Category.fromJson(json['category']) : null,
    );
  }

  Map<String, dynamic> toJson() => {
    'id': id,
    'category_id': categoryId,
    'name': name,
    'description': description,
    'price': price,
    'stock': stock,
    'is_active': isActive,
    'image': image,
    'category': category?.toJson(),
  };
}

File: backend/app/Models/Product.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;

class Product extends Model
{
    protected $fillable = ['category_id', 'name', 'slug', 'description', 'sku', 'price', 'hpp', 'stock', 'image', 'is_active'];

    protected $casts = [
        'price' => 'decimal:2',
        'hpp' => 'decimal:2',
    ];

    protected $appends = ['profit'];

    public function category()
    {
        return $this->belongsTo(Category::class);
    }

    public function ingredients()
    {
        return $this->hasMany(ProductIngredient::class);
    }

    /**
     * Calculate HPP from ingredients and save to the database.
     */
    public function calculateHpp(): float
    {
        $this->load('ingredients.rawMaterial');
        $hpp = 0;

        foreach ($this->ingredients as $ingredient) {
            if ($ingredient->rawMaterial) {
                $hpp += $ingredient->quantity_used * (float) $ingredient->rawMaterial->price_per_small_unit;
            }
        }

        $this->updateQuietly(['hpp' => round($hpp, 2)]);
        return $hpp;
    }

    /**
     * Calculate max possible stock from ingredients and save to database.
     */
    public function syncStock(): int
    {
        $this->load('ingredients.rawMaterial');
        $maxServings = null;

        if ($this->ingredients->isEmpty()) {
            return (int) $this->stock; 
        }

        foreach ($this->ingredients as $ingredient) {
            if (!$ingredient->rawMaterial) continue;

            $available = (float) $ingredient->rawMaterial->stock;
            $needed = (float) $ingredient->quantity_used;

            if ($needed > 0) {
                $servings = floor($available / $needed);
                if ($maxServings === null || $servings < $maxServings) {
                    $maxServings = $servings;
                }
            }
        }

        $newStock = $maxServings === null ? 0 : (int) $maxServings;
        
        if ($this->stock != $newStock) {
            // updateQuietly prevents firing 'saved' again if we had observers
            $this->updateQuietly(['stock' => $newStock]);
        }

        return $newStock;
    }

    /**
     * Profit = selling price - HPP
     */
    public function getProfitAttribute(): float
    {
        return (float) $this->price - (float) $this->hpp;
    }
}
